Ladybird/AppKit/Application/EventLoopImplementation.mm

```diff
@@ -0,0 +1,197 @@
+
+    return false;
+}
+
+static void socket_notifier(CFSocketRef, CFSocketCallBackType, CFDataRef, void const*, void* info)
+{
+    auto& notifier = *reinterpret_cast<Core::Notifier*>(info);
+
+    Core::NotifierActivationEvent event(notifier.fd());
+    notifier.dispatch_event(event);
+}
+
+void CFEventLoopManager::register_notifier(Core::Notifier& notifier)
+{
+    auto notification_type = kCFSocketNoCallBack;
+
+    switch (notifier.type()) {
+    case Core::Notifier::Type::Read:
+        notification_type = kCFSocketReadCallBack;
+        break;
+    case Core::Notifier::Type::Write:
+        notification_type = kCFSocketWriteCallBack;
+        break;
+    default:
+        TODO();
+        break;
+    }
+
+    CFSocketContext context { .info = &notifier };
+    auto socket = CFSocketCreateWithNative(kCFAllocatorDefault, notifier.fd(), notification_type, &socket_notifier, &context);
+
+    auto source = CFSocketCreateRunLoopSource(kCFAllocatorDefault, socket, 0);
+    CFRunLoopAddSource(CFRunLoopGetCurrent(), source, kCFRunLoopDefaultMode);
+    CFRelease(source);
+
+    ThreadData::the().notifiers.set(&notifier, source);
+}
+
+void CFEventLoopManager::unregister_notifier(Core::Notifier& notifier)
+{
+    if (auto source = ThreadData::the().notifiers.take(&notifier); source.has_value()) {
+        CFRunLoopRemoveSource(CFRunLoopGetCurrent(), *source, kCFRunLoopDefaultMode);
+        CFRelease(*source);
+    }
+}
+
+void CFEventLoopManager::did_post_event()
+{
+    auto* event = [NSEvent otherEventWithType:NSEventTypeApplicationDefined
+                                     location:NSMakePoint(0, 0)
+                                modifierFlags:0
+                                    timestamp:0
+                                 windowNumber:0
+                                      context:nil
+                                      subtype:0
+                                        data1:0
+                                        data2:0];
+
+    [NSApp postEvent:event atStart:NO];
+}
+
+NonnullOwnPtr<CFEventLoopImplementation> CFEventLoopImplementation::create()
+{
+    return adopt_own(*new CFEventLoopImplementation);
+}
+
+int CFEventLoopImplementation::exec()
+{
+    [NSApp run];
+    return m_exit_code;
+}
+
+size_t CFEventLoopImplementation::pump(PumpMode mode)
+{
+    auto* wait_until = mode == PumpMode::WaitForEvents ? [NSDate distantFuture] : [NSDate distantPast];
+
+    auto* event = [NSApp nextEventMatchingMask:NSEventMaskAny
+                                     untilDate:wait_until
+                                        inMode:NSDefaultRunLoopMode
+                                       dequeue:YES];
+
+    if (event) {
+        if (event.type == NSEventTypeApplicationDefined) {
+            m_thread_event_queue.process();
+        } else {
+            [NSApp sendEvent:event];
+        }
+    }
+
+    return 0;
+}
+
+void CFEventLoopImplementation::quit(int exit_code)
+{
+    m_exit_code = exit_code;
+    [NSApp stop:nil];
+}
```

Comment:
if the NSApp quits on its own via e.g. The user closing the window, or some other CoreFoundation event from the system, will that bubble up to the return value from main() or from the Core::EventLoop? Is that something we need to worry about? (nested event loops maybe?)

Reply:
It doesn't bubble up, and I don't know if we need to worry about that 🙃  cmd+q and other closed-from-CF events just exit the program, `main` doesn't get to return an exit code.

Reply:
Is there not a "CF Event Loop exited" event that doesn't close the app? You're allowed to have multiple event loops, so I would expect that to be a thing

Reply:
There's `[NSApplicationDelegate applicationWillTerminate]`, but even if I quit the `Core::EventLoop` there, `main` doesn't continue to run.

Reply:
Hm maybe I'm just misunderstanding what this is a wrapper around. Why don't we have a CFRunLoopRun/CFRunLoopStop pair for each of these CFEventLoopImplementation?

Reply:
.. I guess that's what [NSApp run] and [NSApp stop] are doing?

Reply:
There's also a `CFRunLoopObserverCreateWithHandler`, where we can attach an observer to the run loop for `kCFRunLoopExit` notifications, but I'm not seeing it get fired on exit.

Reply:
Right, `NSApp` wraps around the CF run loop, which we have to use for the GUI to show at all.

Reply:
I added an `NSApplication` subclass to override `terminate:`, so the implementation can quit the app via `Core::EventLoop::quit`. We exit by returning from `main` now.

I also added a note that this event loop implementation is only for wrapping around the `NSApp` event loop, if we need multiple event loops in the future, we'll have to start and run our own `CFRunLoop`.